src/Tar.php

```diff
@@ -573,12 +573,19 @@ protected function parseHeader($block)
         // Handle Long-Link entries from GNU Tar
         if ($return['typeflag'] == 'L') {
             // following data block(s) is the filename
-            $filename = trim($this->readbytes(ceil($header['size'] / 512) * 512));
+            $filename = trim($this->readbytes(ceil($return['size'] / 512) * 512));
             // next block is the real header
             $block  = $this->readbytes(512);
             $return = $this->parseHeader($block);
+
             // overwrite the filename
             $return['filename'] = $filename;
+        }elseif ($return['typeflag'] == 'x') {
+            // following data block(s) is the filename
+            $filename = trim($this->readbytes(ceil($return['size'] / 512) * 512));
+            // next block is the real header
+            $block  = $this->readbytes(512);
+            $return = $this->parseHeader($block);
```

Comment:
I don't understand. You are not using the data read from `x` header. $filename is never used. You basically just read and ignore the `x` header.

Since your tests work, I assume the original header already contains the proper UTF-8 filename? Would a very long UTF-8 filename have an `x` and an `L` header?